📖 Document the immutability of machine templates (#3105)
kubernetes-sigs/cluster-api: docs/book/src/tasks/change-machine-template.md

# Changing Infrastructure Machine Templates

Several different components of Cluster API leverage _infrastructure machine templates_, 
including `KubeadmControlPlane`, `MachineDeployment`, and `MachineSet`. These 

> `MachineTemplate` resources are **immutable** and are not intended to be modified 

Reviewer: xref https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/1711. I know you mention at the bottom that providers may decide to support some field changes, but I wonder if we should highlight this use case here, or soften/clarify the **immutable** bit here?

Reviewer 2: +1 I'd say that they _should_ be immutable, unless the documentation on the infrastructure provider states otherwise for certain fields

cc @rudoi @sethp-nr

Author: Addressed via af55bb4.

directly.

The correct process for modifying an infrastructure machine template is as follows:

1. Duplicate an existing template.
    Users can use `kubectl get <MachineTemplateType> <name> -o yaml > file.yaml` 
    to retrieve a template configuration from a running cluster to serve as a starting 
    point.
2. Update the desired fields.
    Fields that might need to be modified could include the SSH key, the AWS instance 

>     type, or the Azure machine type). Refer to the provider-specific documentation 

Reviewer 3: ```suggestion
    type, or the Azure VM size). Refer to the provider-specific documentation 
```

Author: Changed via af55bb4.

    for more details on the specific fields that each provider requires or accepts.

> 3. Give the newly-modified template a new name by modifying the `metadata.name` field.

Reviewer 2: Or using metadata.generateName?

Author: Added via af55bb4.

4. Create the new infrastructure machine template on the API server using `kubectl`.

Once the new infrastructure machine template has been persisted, users may modify 
the object that was referencing the infrastructure machine template. For example, 
to modify the infrastructure machine template for the `KubeadmControlPlane` object, 
users would modify the `spec.infrastructureTemplate.name` field. For a `MachineDeployment` 
or `MachineSet`, users would need to modify the `spec.template.spec.infrastructureRef.name` 

> field. In all cases, the `name` field should be update to point to the newly-modified 

Reviewer: ```suggestion
field. In all cases, the `name` field should be updated to point to the newly-modified 
```

Author: Fixed via af55bb4.

infrastructure machine template. This will trigger a rolling update. (This same process 
is described in the documentation for [upgrading the underlying machine image for 

> KubeadmControlPlane](./kubeadm-control-plane.html#how-to-upgrade-the-underlying-machine-image).)

Reviewer: I don't think CI likes this link

Author: The link worked when I ran `mdbook` locally, but I've simplified the link in af55bb4 and hopefully this will work now.

# Changing Infrastructure Machine Templates

Several different components of Cluster API leverage _infrastructure machine templates_, 
including `KubeadmControlPlane`, `MachineDeployment`, and `MachineSet`. These 
`MachineTemplate` resources should be immutable, unless the infrastructure provider 
documentation indicates otherwise for certain fields (see below for more details).

The correct process for modifying an infrastructure machine template is as follows:

1. Duplicate an existing template.

>     Users can use `kubectl get <MachineTemplateType> <name> -o yaml > file.yaml` 

Reviewer: Not sure if you're trying to have this formatted as-written in here, but the whitespace & indentation are lost when rendered. See https://deploy-preview-3105--kubernetes-sigs-cluster-api.netlify.app/tasks/change-machine-template.html.

Author: I'm OK with the rendered version if you are. That wasn't _quite_ what I was shooting for, but it's acceptable IMO.

    to retrieve a template configuration from a running cluster to serve as a starting 
    point.
2. Update the desired fields.
    Fields that might need to be modified could include the SSH key, the AWS instance 

>     type, or the Azure VM size). Refer to the provider-specific documentation 

Reviewer: ```suggestion
    type, or the Azure VM size. Refer to the provider-specific documentation 
```

Author: Fixed via 5810f14

    for more details on the specific fields that each provider requires or accepts.
3. Give the newly-modified template a new name by modifying the `metadata.name` field 
    (or by using `metadata.generateName`).

> 4. Create the new infrastructure machine template on the API server using `kubectl`.

Reviewer: If you saved to disk as indicated in step 1, you'll also need to clear out the resource version before trying to send it to the apiserver.

Author: @ncdc Should I omit this command entirely? Is there a better way to generate a new machine template?

Reviewer: I would just add a parenthetical that says make sure you clear out the RV

Author: @ncdc Have a look at this updated wording and let me know what you think (added in 276fe9f).
